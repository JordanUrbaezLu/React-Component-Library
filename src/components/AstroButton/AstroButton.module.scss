.astroButton {
  border-radius: 30px;
  width: fit-content;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
  &:hover {
    cursor: pointer;
  }
  display: flex;
  flex-direction: row;
  align-items: center;
  column-gap: 5px;
}

.primary {
  background-color: rgb(255, 183, 77);
  color: black;
  border: 2px solid rgb(255, 183, 77);
  &:hover {
    color: rgb(147 51 234);
    border: 2px solid rgb(147 51 234);
  }
  &:active {
    background: rgb(255, 183, 77, 0.8);
  }
}

.secondary {
  background-color: rgb(147 51 234);
  color: black;
  border: 2px solid rgb(147 51 234);
  &:hover {
    color: rgb(255, 183, 77);
    border: 2px solid rgb(255, 183, 77);
  }
  &:active {
    background: rgb(147, 51, 234, 0.8);
  }
}

.small {
  font-size: 12px;
  padding: 6px 8px;
  .leadingIcon,
  .trailingIcon {
    height: 12px;
  }
}

.medium {
  font-size: 16px;
  padding: 8px 10px;
  .leadingIcon,
  .trailingIcon {
    height: 16px;
  }
}

.large {
  font-size: 20px;
  padding: 10px 12px;
  .leadingIcon,
  .trailingIcon {
    height: 20px;
  }
}

.disabled {
  color: gray !important;
  border: 2px solid gray !important;
  background: rgb(0, 0, 0, 0.1) !important;
  cursor: default !important;
}
