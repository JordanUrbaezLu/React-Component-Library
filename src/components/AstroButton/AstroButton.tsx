import * as React from "react";
import styles from "./AstroButton.module.scss";
import classNames from "classnames";

export type AstroButtonVariants = "primary" | "secondary";
export type AstroButtonSize = "small" | "medium" | "large";
export type AstroButtonType = "button" | "submit" | "reset";

interface AstroButtonProps {
  leadingIcon?: React.ReactNode;
  trailingIcon?: React.ReactNode;
  size?: AstroButtonSize;
  variant?: AstroButtonVariants;
  children: React.ReactNode;
  disabled?: boolean;
  type?: AstroButtonType; //TODO add tests to type?
  onClick?: () => void;
}

const AstroButton = React.forwardRef<HTMLButtonElement, AstroButtonProps>(
  (props, ref) => {
    const classes = classNames(
      styles.astroButton,
      props.size === "small"
        ? styles.small
        : props.size === "large"
        ? styles.large
        : styles.medium,
      props.variant === "secondary" ? styles.secondary : styles.primary,
      props.disabled && styles.disabled
    );

    const content = (
      <>
        {props.leadingIcon && (
          <span className={styles.leadingIcon}>{props.leadingIcon}</span>
        )}
        {props.children}
        {props.trailingIcon && (
          <span className={styles.trailingIcon}>{props.trailingIcon}</span>
        )}
      </>
    );

    return (
      <button
        type={
          props.type === "reset"
            ? "reset"
            : props.type === "submit"
            ? "submit"
            : "button"
        }
        onClick={props.onClick}
        ref={ref}
        className={classes}
      >
        {content}
      </button>
    );
  }
);

export default AstroButton;
