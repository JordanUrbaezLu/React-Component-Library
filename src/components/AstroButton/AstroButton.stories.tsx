import { ComponentStory, ComponentMeta } from "@storybook/react";
import AstroButton from "./AstroButton";
import { AiOutlineArrowLeft, AiOutlineArrowRight } from "react-icons/ai";

export default {
  title: "AstroButton",
  component: AstroButton,
} as ComponentMeta<typeof AstroButton>;

const Template: ComponentStory<typeof AstroButton> = (args) => (
  <AstroButton {...args}>Button</AstroButton>
);

export const Default = Template.bind({});

export const Primary = Template.bind({});
Primary.args = {
  variant: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const LeadingIcon = Template.bind({});
LeadingIcon.args = {
    leadingIcon: <AiOutlineArrowLeft />,
}

export const TrailingIcon = Template.bind({});
TrailingIcon.args = {
    trailingIcon: <AiOutlineArrowRight />,
}

export const LeadingIconAndTrailingIcon = Template.bind({});
LeadingIconAndTrailingIcon.args = {
    leadingIcon: <AiOutlineArrowLeft />,
    trailingIcon: <AiOutlineArrowRight />
}

