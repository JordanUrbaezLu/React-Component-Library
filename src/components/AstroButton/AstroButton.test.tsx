import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import AstroButton, {
  AstroButtonSize,
  AstroButtonVariants,
} from "./AstroButton";
import { axe, toHaveNoViolations } from "jest-axe";
expect.extend(toHaveNoViolations);

describe("AstroButton", () => {
  test("Should render component correctly.", () => {
    const { container } = render(<AstroButton>AstroButton</AstroButton>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(<AstroButton>AstroButton</AstroButton>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test.each<AstroButtonVariants>(["primary", "secondary"])(
    "Should render variants correctly.",
    (variant) => {
      render(<AstroButton variant={variant}>AstroButton</AstroButton>);
      expect(screen.getByRole("button", { name: "AstroButton" })).toHaveClass(
        variant
      );
    }
  );
  test.each<AstroButtonSize>(["small", "medium", "large"])(
    "Should render sizes correctly",
    (size) => {
      render(<AstroButton size={size}>AstroButton</AstroButton>);
      expect(screen.getByRole("button", { name: "AstroButton" })).toHaveClass(
        size
      );
    }
  );
  test("Should render leading icon correctly.", () => {
    render(<AstroButton leadingIcon="leading">AstroButton</AstroButton>);
    expect(
      screen.getByRole("button", { name: "leading AstroButton" })
    ).toBeInTheDocument();
  });

  test("Should render trailing icon correctly.", () => {
    render(<AstroButton trailingIcon="trailing">AstroButton</AstroButton>);
    expect(
      screen.getByRole("button", { name: "AstroButton trailing" })
    ).toBeInTheDocument();
  });

  test("Should render disabled correctly.", () => {
    render(<AstroButton disabled>AstroButton</AstroButton>);
    expect(screen.getByRole("button", { name: "AstroButton" })).toHaveClass(
      "disabled"
    );
  });

  test("Should render children correctly.", () => {
    render(<AstroButton>AstroButton</AstroButton>);
    expect(
      screen.getByRole("button", { name: "AstroButton" })
    ).toBeInTheDocument();
  });
});
