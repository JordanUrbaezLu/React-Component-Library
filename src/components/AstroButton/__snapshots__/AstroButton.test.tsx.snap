// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroButton Should render component correctly. 1`] = `
<button
  class="astroButton medium primary"
  type="button"
>
  AstroButton
</button>
`;
